src/wallet/wallet.h

```diff
@@ -1544,5 +1542,162 @@ class AddSpendingKeyToWallet : public boost::static_visitor<KeyAddResult>
+            }
+        }
+        else {
+            LogPrint("balanceunsafe", "Balance changed in address %s from %s to %s\n",
+                address, ValueFromAmount(0), ValueFromAmount(amount));
+            balances[address] = amount;
+        }
+    }
+
+    TransactionStateType LogState(const uint256 &hashTx)
+    {
+        TransactionStateType state;
+        if (txstate.count(hashTx) == 0) {
+            state = TX_GENERATED;
+            txstate.insert(std::make_pair(hashTx, state));
+            LogPrint("stateunsafe", "Transaction generated: %s\n", hashTx.ToString());
+        }
+        else {
+            if(pwalletMain->mapWallet[hashTx].IsInMainChain()) {
+                state = TX_MINED;
+                txstate[hashTx] = state;
+                LogPrint("stateunsafe", "Transaction mined: %s\n", hashTx.ToString());
+            }
+        }
+        return state;
+    }
+
+    void operator()(const uint256 &hashTx)
+    {
+        const auto& tx = pwalletMain->mapWallet[hashTx];
+        TransactionStateType state;
+        bool doStateLog = true; // used to only log state once per tx
+
+        // sapling notes
+        std::set<libzcash::SaplingPaymentAddress> saplingAdresses;
+        pwalletMain->GetSaplingPaymentAddresses(saplingAdresses);
+
+        for (auto& note : tx.mapSaplingNoteData) {
+            const SaplingOutPoint& op = note.first;
+            const SaplingNoteData& nd = note.second;
+
+            const auto maybe_pt = libzcash::SaplingNotePlaintext::decrypt(
+                tx.vShieldedOutput[op.n].encCiphertext,
+                nd.ivk,
+                tx.vShieldedOutput[op.n].ephemeralKey,
+                tx.vShieldedOutput[op.n].cmu);
+
+            assert(static_cast<bool>(maybe_pt));
+            const auto& notePt = maybe_pt.get();
+
+            const auto& maybe_pa = nd.ivk.address(notePt.d);
+            assert(static_cast<bool>(maybe_pa));
+            const auto& pa = maybe_pa.get();
+
+            if (saplingAdresses.count(pa) > 0) {
+                const auto& notevalue = notePt.note(nd.ivk).get();
+                if (doStateLog) {
+                    state = LogState(hashTx);
+                    doStateLog = false;
+                }
+                if (state == TX_MINED)
+                    LogBalance(EncodePaymentAddress(pa), notevalue.value(), false);
+            }
+        }
+
+        // sprout notes
+        std::set<libzcash::SproutPaymentAddress> sproutAdresses;
+        pwalletMain->GetSproutPaymentAddresses(sproutAdresses);
+
+        for (auto& note : tx.mapSproutNoteData) {
+            const JSOutPoint& jsop = note.first;
+            const SproutNoteData& nd = note.second;
+            const auto decrypted = tx.DecryptSproutNote(jsop);
+
+            if (HaveSpendingKeyForPaymentAddress(pwalletMain)(decrypted.second)) {
+                const auto& address = EncodePaymentAddress(decrypted.second);
+                if (sproutAdresses.count(decrypted.second) > 0) {
```

Comment:
Use the `PaymentAddressBelongsToWallet` visitor instead of these nested checks.

Reply:
Similarly to above, the checks can actually be removed.

Reply:
https://github.com/zcash/zcash/pull/4423/commits/351efb418ac341745ca46278eb92fff6749d5519